SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,615 @@
+{
+	Worker_EntityId Entity1 = Entities[Slot1];
+	Worker_EntityId Entity2 = Entities[Slot2];
+
+	SlotMap[Entity1] = Slot2;
+	SlotMap[Entity2] = Slot1;
+
+	Swap(Entities[Slot1], Entities[Slot2]);
+	Swap(EntityPosition[Slot1], EntityPosition[Slot2]);
+	Swap(EntityPositionX[Slot1], EntityPositionX[Slot2]);
+	Swap(EntityPositionY[Slot1], EntityPositionY[Slot2]);
+	Swap(EntityFlags[Slot1], EntityFlags[Slot2]);
+	Swap(ActiveTimestamp[Slot1], ActiveTimestamp[Slot2]);
+	Swap(Visibility[Slot1], Visibility[Slot2]);
+
+	if (Broadphase)
+	{
+		Broadphase->SwapEntries(Slot1, Slot2);
+	}
+}
+
+void FInterestManager::ComputeInterest(ISpatialOSWorker& Connection, const TArray<Worker_EntityId> Workers, const TArray<FBox2D>& Regions)
+{
+	SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputation);
+
+	if (Entities.Num() == 0)
+	{
+		return;
+	}
+	if (Regions.Num() == 0)
+	{
+		return;
+	}
+
+	// Set an entity that has not moved since 5 secs as inactive. It will not be checked against the regions.
+	// This assumes that the bits allocation against workers remains constant, which is not valid anymore when changing/reassigning regions.
+	const double InactiveTime = 5.0;
+	const uint64 TimeToConsiderInactive = FPlatformTime::Cycles64() - (InactiveTime / FPlatformTime::GetSecondsPerCycle64());
+	for (int32 i = 0; i < (int32)ActiveEntities; ++i)
+	{
+		if (ActiveTimestamp[i] < TimeToConsiderInactive)
+		{
+			SwapEntries(i, ActiveEntities - 1);
+			--ActiveEntities;
+			--i;
+		}
+	}
+
+	check(Regions.Num() <= 64);
+
+	const uint32 NumRegions = Regions.Num();
+	const uint32 NumEntities = Entities.Num();
+
+	if (CachedServerInterest[0].Num() < (int32)NumRegions)
+	{
+		CachedServerInterest[0].SetNum(NumRegions);
+		CachedServerInterest[1].SetNum(NumRegions);
+		CachedServerInterest[2].SetNum(NumRegions);
+	}
+
+	constexpr bool bUseBroadphaseForRegions = false;
+
+	if (Broadphase && bUseBroadphaseForRegions)
+	{
+		// 60ms, 200k entities, 64 regions, 5% overlap
+		// Worse than the non-spatialized version, which is not surprising, since we are visiting all the entities anyway,
+		// hence visiting the tree before testing the entities is just more data to process.
+		// Have to test it for the player interest since the tree culling might actually do something in that case.
+
+		// Since the regions are a partition of the entire space in the server case, a better strategy would be to find the region in the
+		// partition an entity belongs to and test the position against the segment boundaries with the interest border. The region's
+		// connectivity would be needed to interpret these results.
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBroadphase);
+		const uint32 NumActiveEntities = ActiveEntities;
+
+		uint64* VisibilityPtr = Visibility.GetData();
+		const uint32* Flags = EntityFlags.GetData();
+		for (uint32 i = 0; i < NumActiveEntities; ++i)
+		{
+			const uint64 Mask = UINT64_MAX;
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * Mask;
```

Comment:
worth a comment explaining what this masking operation achieves?

Reply:
So the value at each array entry is either UINT64_MAX or 0 depending on whether Flags has any non 0 value?

Reply:
added comment